src/inmanta/server/services/orchestrationservice.py

```diff
@@ -1032,87 +1032,112 @@ async def release_version(
-            # not checking error conditions
-            await self.resource_service.resource_action_update(
-                env,
-                undep_ids,
-                action_id=uuid.uuid4(),
-                started=now,
-                finished=now,
-                status=const.ResourceState.undefined,
-                action=const.ResourceAction.deploy,
-                changes={},
-                messages=[],
-                change=const.Change.nochange,
-                send_events=False,
-                connection=connection,
-            )
-
-            skippable = await model.get_skipped_for_undeployable()
-            if skippable:
-                skippable_ids = [ResourceVersionIdStr(rid + ",v=%s" % version_id) for rid in skippable]
-                # not checking error conditions
-                await self.resource_service.resource_action_update(
-                    env,
-                    skippable_ids,
-                    action_id=uuid.uuid4(),
-                    started=now,
-                    finished=now,
-                    status=const.ResourceState.skipped_for_undefined,
-                    action=const.ResourceAction.deploy,
-                    changes={},
-                    messages=[],
-                    change=const.Change.nochange,
-                    send_events=False,
-                    connection=connection,
+        async with data.ConfigurationModel.get_connection(connection) as connection:
+            async with connection.transaction():
+                model = await data.ConfigurationModel.get_version_internal(
+                    env.id, version_id, connection=connection, lock=RowLockMode.FOR_UPDATE
                 )
+                if model is None:
+                    return 404, {"message": "The request version does not exist."}
 
-        increments: tuple[abc.Set[ResourceIdStr], abc.Set[ResourceIdStr]] = await self.resource_service.get_increment(
-            env, version_id
-        )
+                if model.released:
+                    raise Conflict(f"The version {version_id} on environment {env} is already released.")
 
-        increment_ids, neg_increment = increments
-        await self.resource_service.mark_deployed(env, neg_increment, now, version_id)
+                latest_version = await data.ConfigurationModel.get_version_nr_latest_version(env.id, connection=connection)
 
-        # Set the updated field:
-        await data.Resource.copy_last_success(env.id, version_id)
+                # ensure we are the latest version
+                # this is required for the subsequent increment calculation to make sense
+                # this does introduce a race condition, with any OTHER release running concurrently on this environment
+                # We could lock the get_version_nr_latest_version for update to prevent this
+                if model.version < (latest_version or -1):
+                    raise Conflict(
```

Comment:
This check didn't exist either, it is still a bit of a stink like this, but I'm a bit uncertain on how to test it if I would add another lock here.

Reply:
yes. This one is safe wrt itself